### PR TITLE
Add `parse_uri()` / `parse_url()` function

### LINKED ISSUE #11941: Add URL data type
**Use case**: It would be *really useful* to have a URL data type that would verify that a URL is correctly formatted when being inserted and more importantly let you do operations like GROUP BY url(path, url_field_name) efficiently.

**Feature description**: Validate URL upon insertion / update. Automatically store the URL in a way so that the parts of the URL are easily accessible and indexed efficiently. Turning the querystring into key/value pairs that you could easily search would be immensely helpful.

Parts of the URL that should be indexed:

- scheme
- hostname
- domain
- port
- path
- querystring parameters (take ?this=that&color=blue&validate should turn into key value pairs that can be searched with a WHERE clause easily or used in a GROUP BY)
- hash fragment

Lastly it would be great to automatically decode any [percent encoded](https://en.wikipedia.org/wiki/Percent-encoding) values.

Presently I've had to split up URLs into a bunch of TEXT fields when storing them in Crate and it seems like not a great way to do things.

### PR DESCRIPTION
**Pull request by proddata**
## Summary of the changes / Why this improves CrateDB

Adding a `parse_uri(<TEXT>)` and `parse_url(<TEXT>)` function, that could resolve #11941.

**References from other DBs**
- `parse_url()` from [Snowflake](https://docs.snowflake.com/en/sql-reference/functions/parse_url.html)
- [url functions from ClickHouse](https://clickhouse.com/docs/en/sql-reference/functions/url-functions/)
- `parse_url()` from [Databricks](https://docs.databricks.com/sql/language-manual/functions/parse_url.html)

---

```sql
cr> SELECT parse_uri('https://testing.crate.io:4200/') as url;
+--------------------------------------------------------------------------------+
| url                                                                            |
+--------------------------------------------------------------------------------+
| {"hostname": "testing.crate.io", "path": "/", "port": 4200, "scheme": "https"} |
+--------------------------------------------------------------------------------+
SELECT 1 row in set (0.003 sec)
```

```sql
cr> SELECT parse_url('https://user:pw%26@testing.crate.io:4200/sub+space/sub%20space2/index.html?foo=bar&foo=&foo2=https%3A%2F%2Fcrate.io%2F%3Ffoo%3Dbar%26foo%3Dbar2%26foo2#ref') as url;              
+---------------------------------------------------------------------------+
|  url                                                                      |
+---------------------------------------------------------------------------+
| {                                                                         |
|   "fragment": "ref",                                                      |
|   "hostname": "testing.crate.io",                                         |
|   "parameters":                                                           |
|     {                                                                     |
|       "foo": ["bar", null],                                               |
|       "foo2": ["https://crate.io/?foo=bar&foo=bar2&foo2"],                |
|     },                                                                    |
|   "path": "/sub space/sub space2/index.html",                             |
|   "port": 4200,                                                           |
|   "query": "foo=bar&foo=&foo2=https://crate.io/?foo=bar&foo=bar2&foo2",   |
|   "scheme": "https",                                                      |
|   "userinfo": "user:pw&"                                                  |
| }                                                                         |
+---------------------------------------------------------------------------+
```

```sql
CREATE TABLE uri_table (
   url TEXT,
   url_object OBJECT GENERATED ALWAYS AS parse_uri("url")
);
```

This would be currently rather hard to implement in CrateDB with UDFs as the GraalVM setup doesn't include the URL API

cc: @robd003 

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by proddata** [comment]
@seut / @jweberpm feedback appreciated

**Comment by robd003** [comment]
I've tried it on both 5.0-SNAPSHOT builds and even cherry-picking this to the 4.8 branch and both are working well!

**Comment by jweberpm** [comment]
> @seut / @jweberpm feedback appreciated

Looks good from all I can see and judge, ok still missing changelog/docs but otherwise looks good for me.

File: server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                                                             uriText));
+            }
+        }
+
+        putNonNullValue(uriMap,"scheme", scheme);
+        putNonNullValue(uriMap,"hostname", hostname);
+        putNonNullValue(uriMap,"port", port == -1 ? null : port);
+        putNonNullValue(uriMap,"path", decodeText(path));
+        putNonNullValue(uriMap,"query", parseQuery(query));
+        putNonNullValue(uriMap,"fragment", fragment);
```
Comment (proddata) [inline server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java:137]: probably should also be decoded

File: server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+public final class ParseURIFunction extends Scalar<Object, String> {
+
+    private static final String NAME = "parse_uri";
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.UNTYPED_OBJECT.getTypeSignature()
```
Comment (seut) [inline server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java:57]: I suggest to use a typed object with fixed inner elements which can have a `NULL` value. This way the expected returned object has a strict schema, which would help when working with it. If some keys aren't available, one would have to check their existence.

File: server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+        String scheme;
+        String hostname;
+        Integer port;
+        String path;
+        String query;
+        String fragment;
+        String userinfo;
+
+        // Unfortunately the URI class is rather strict in its processing and fails for non URI-encoded characters
```
Comment (seut) [inline server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java:105]: As `URL` defines a subset of `URI` I wonder in which cases parsing it as an URI fails but succeeds with URL. If it contains non URI-encoded characters, it's not a valid URI and this method should fail.

Comment (seut) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java:105]: > I'd rather tend to be more flexibel, as the parsed Object is indeed valid in both cases. A parse_text(<URL>) method imho should follow RFC2396

Shouldn't this method then be called `parse_url` instead? and thus also supports URL's.
Anyhow, is there really any use-case to support parsing of non-valid-encoded URL's?

**Comment by robd003** [comment]
Any chance we could get this merged into main before the 5.0 release is cut? @seut / @mfussenegger

**Comment by proddata** [comment]
> Any chance we could get this merged into main before the 5.0 release is cut? @seut / @mfussenegger

I think this would need some rework (i.e. split up logic between `URI` and `URL`).
i.e the query parsing is rather `URL` specific and would probably fail for some URIs

**Comment by mfussenegger** [comment]
> Any chance we could get this merged into main before the 5.0 release is cut? @seut / @mfussenegger

Unfortunately not. We already kinda hit the feature freeze cut and as @proddata pointed out the PR needs some more work.

File: server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java (hunk 3 above)
Comment (proddata) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java]: I split the functions into 2 separate ones. While URLs are URIs, URLs are more specific about how e.g. the query string can be formatted. Overall the `parse_url` functions imho has a higher priority.

**Comment by robd003** [comment]
Thanks so much for this @proddata, Happy New Year!

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        return urlMap;
+    }
+
+    private final Map<String,List<String>> parseQuery(String query) {
+        if (Strings.isNullOrEmpty(query)) {
+            return null;
+        }
+
+        Map<String,List<String>> queryMap = new HashMap<String,List<String>>();
+        Arrays.stream(query.split("&(?!amp)"))
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java:122]: with smth like

```
String[] params = query.split("&(?!amp)")
for (String param: params)...
```
we could avoid overhead of creating a stream

File: server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    public final Object evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>... args) {
+        String uri = args[0].value();
+        if (uri == null) {
+            return null;
+        }
+        return parseURI(uri);
+    }
+
+    private final Object parseURI(String uriText) {
+        final Map<String, Object> uriMap = new LinkedHashMap<>();
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java:83]: Just curious - is order of the `URI` parts important? Otherwise could use regular `HashMap` to keep it simple.

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    public final Object evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>... args) {
+        String url = args[0].value();
+        if (url == null) {
+            return null;
+        }
+        return parseURL(url);
+    }
+
+    private final Object parseURL(String urlText) {
+        final Map<String, Object> urlMap = new LinkedHashMap<>();
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java:92]: ditto

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    private final void parseQueryParameters(Map<String,List<String>> map, String params) {
+        final int idx = params.indexOf("=");
+        final String key = idx > 0 ? decodeText(params.substring(0, idx)) : decodeText(params);
+        final String value = idx > 0 && params.length() > idx + 1 ? decodeText(params.substring(idx + 1)) : null;
+        if (!map.containsKey(key)) {
+            map.put(key, new ArrayList<String>());
+        }
+        map.get(key).add(value);
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java:135]: I think map.computeIfAbsent(key, k -> new ArrayList<>()).add(value); would work here - less verbose and saves one lookup if key exits.

Regarding `ArrayList` - maybe can be `Set` depending on what we want. For query string like `key=val1&another_key=val&key=val1`, we end up having key-> (val, val). It's fine but we could drop the duplicates. If users want original query string, they can use "query" field but parse_url("someurl")[parameters] can have unique vals. No strong opinion here.

File: server/src/test/java/io/crate/expression/scalar/ParseURIFunctionTest.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        Map<String, Object> value = MapBuilder.<String, Object>newMapBuilder()
+            .put("scheme", "https")
+            .put("userinfo", null)
+            .put("hostname", "crate.io")
+            .put("port", null)
+            .put("path", "/index.html")
+            .put("query", null)
+            .put("fragment", null)
+            .map();
+        assertEvaluate(String.format("parse_uri('%s')", uri), value);
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/expression/scalar/ParseURIFunctionTest.java:49]: let's use `Locale.ENGLISH` for formatting to be consistent with the codebase

File: server/src/test/java/io/crate/expression/scalar/ParseURLFunctionTest.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        Map<String, Object> value = MapBuilder.<String, Object>newMapBuilder()
+            .put("scheme", "https")
+            .put("userinfo", "user:pw&")
+            .put("hostname", "testing.crate.io")
+            .put("port", 4200)
+            .put("path", "/sub space/sub space2/index.html")
+            .put("query","foo=bar&foo=&foo2=https://crate.io/?foo=bar&foo=bar2&foo2")
+            .put("parameters", MapBuilder.<String,Object>newMapBuilder()
+                                .put("foo",Stream.of("bar",null).toList())
+                                .put("foo2",List.of("https://crate.io/?foo=bar&foo=bar2&foo2"))
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/expression/scalar/ParseURLFunctionTest.java:118]: Parameter containing `&` is an interesting edge case - maybe worth adding such example to the docs?
Let's even make it harder and and a third `key=val` after `param_containing_amp` (here and in docs)

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        Map<String,List<String>> queryMap = new HashMap<String,List<String>>();
+        Arrays.stream(query.split("&(?!amp)"))
+                     .forEach((String param) -> parseQueryParameters(queryMap, param));
+
+        return queryMap;
+    }
+
+    private final void parseQueryParameters(Map<String,List<String>> map, String params) {
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java:128]: Can we rename `params` to `paramAndVal` or something bearing such meaning? 
params here is a `key=val` pair so was bit confused when saw plural.

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        final int idx = params.indexOf("=");
+        final String key = idx > 0 ? decodeText(params.substring(0, idx)) : decodeText(params);
+        final String value = idx > 0 && params.length() > idx + 1 ? decodeText(params.substring(idx + 1)) : null;
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java:131]: Not sure but I think values can be a comma separated list, ie, we can pass to `params` string looking like 
`key = val1, val2` which it still fine to treat as a single value even in object. 

We could mention explicitly in docs that comma separated list is NOT converted into array in the `parsed_url_obj[params]` and and a test case. 

Or we can convert comma separated lists to array if you think it's more convenient to users, no strong opinion here.

File: server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java (hunk 5 above)
Comment (proddata) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java:83]: no ... will change it

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java (hunk 7 above)
Comment (proddata) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java:135]: changed to `computeIfAbsent`

I would prefer to use an ArrayList to be able to recreate the original URL, as the `query` string is decoded, which is not always reversible

File: server/src/test/java/io/crate/expression/scalar/ParseURLFunctionTest.java (hunk 9 above)
Comment (proddata) [reply server/src/test/java/io/crate/expression/scalar/ParseURLFunctionTest.java:118]: The `&` in the parameter needs to be encoded. The is standard URL behaviour and not specific to CrateDB. The `query` string just holds the decoded query string, which is in the documentation.

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java (hunk 11 above)
Comment (proddata) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java:131]: Converting to an array is not easy, as it would make it impossible to use a specific type and index it, which is the main purpose of the function.

It is part of the documentation, that the result is an `ARRAY(TEXT)` for each dedicated parameter (key)

Basically one should be able to do the following:

```sql
SELECT * FROM tab WHERE 'hello' = ANY(url['parameters']['my_parameter'])
```

File: server/src/test/java/io/crate/expression/scalar/ParseURIFunctionTest.java (hunk 8 above)
Comment (proddata) [reply server/src/test/java/io/crate/expression/scalar/ParseURIFunctionTest.java]: done

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java (hunk 4 above)
Comment (proddata) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java]: thanks 👍

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java (hunk 10 above)
Comment (proddata) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java]: I decided to move the logic to the outer method, as it doesn't make sense to necessarily separate anymore.

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java (hunk 11 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java]: >It is part of the documentation, that the result is an ARRAY(TEXT)

I saw 
```
"parameters" OBJECT AS (
             "key1" ARRAY(TEXT),
```
However, it might be still not clear from docs (and also not covered in tests) that _each item_ of the `ARRAY` can be a comma separated list. Ie for query string
`key1=single_val&key2=val2&key1=x,y,z` we end up having 
```
parameters {
    key1: ["single_val", "x,y,z"],
    key2:val2
}

```
and NOT
```
parameters {
    key1: ["single_val", "x", "y", "z"],
    key2:val2
}
```

File: server/src/test/java/io/crate/expression/scalar/ParseURLFunctionTest.java (hunk 9 above)
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/expression/scalar/ParseURLFunctionTest.java:119]: Ah ok, so you are saying we won't have ambiguous `split("&"`).

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java (hunk 11 above)
Comment (proddata) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java]: How about adding another example to the docs then?
```sql
cr> SELECT parse_url('http://crate.io?p1=1,2,3&p1=a&p1')['parameters'] as params;                                                                                                                                
+------------------------------+
| params                       |
+------------------------------+
| {"p1": ["1,2,3", "a", null]} |
+------------------------------+
SELECT 1 row in set (0.002 sec)
```

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java]: Sounds good. Maybe add a note before this example, that each item of the array can be a comma separated list

File: server/src/test/java/io/crate/expression/scalar/ParseURLFunctionTest.java (hunk 9 above)
Comment (proddata) [reply server/src/test/java/io/crate/expression/scalar/ParseURLFunctionTest.java:119]: correct. An ampersand in a value has to be escaped as `%26`, otherwise it is separating to keys.

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    private final Map<String,List<String>> parseQuery(String query) {
+        if (Strings.isNullOrEmpty(query)) {
+            return null;
+        }
+
+        Map<String,List<String>> queryMap = new HashMap<String,List<String>>();
+        String[] parameters = query.split("&(?!amp)");
+        for (String parameter : parameters) {
+            final int idx = parameter.indexOf("=");
+            final String key = idx > 0 ? decodeText(parameter.substring(0, idx)) : decodeText(parameter);
```
Comment (seut) [inline server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java:123]: This would not work if the query part doesn't contain a attribute-value pair separated by `=` but just some string, e.g.: `https://foo.com/some_path?bar`
This is a valid query parameter according to the RFC, see https://www.rfc-editor.org/rfc/rfc3986#section-3.4.

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java (hunk 11 above)
Comment (proddata) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java]: Added the example with note

```rst
Parameter values are always treated as ``text``. There is no conversion of 
comma-separated parameter values into arrays::

    cr> SELECT parse_url('http://crate.io?p1=1,2,3&p1=a&p2[]=1,2,3')['parameters'] as params;                                                                                                                        
    +-------------------------------------------+
    | params                                    |
    +-------------------------------------------+
    | {"p1": ["1,2,3", "a"], "p2[]": ["1,2,3"]} |
    +-------------------------------------------+
    SELECT 1 row in set (... sec)
```

---

I also checked, but there is no standard way to handle conversions anyway -> very framework specific.

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java (hunk 12 above)
Comment (proddata) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java:123]: What do you mean with "not work". `bar` is a parameter with `null` as value:

```sql
cr> SELECT parse_url('https://foo.com/some_path?bar')['parameters'] url;                                                                                                                                         
+-----------------+
| url             |
+-----------------+
| {"bar": [null]} |
+-----------------+
SELECT 1 row in set (0.022 sec)
```

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java (hunk 11 above)
Comment (seut) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java]: > However, it might be still not clear from docs (and also not covered in tests) that each item of the ARRAY can be a comma separated list.

According to the URL RFC/Specifications, a query is just defined as an arbitrary string. Even the use of attribute-value pairs is not strongly defined. I would not go such deep and explain well-known behaviour of query parameters in our documentation, thus adding an example about a valid value containing `,` chars is way too much for our docs in my opinion.

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java (hunk 12 above)
Comment (seut) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java:123]: Ah damn, your right and I just missed reading carefully enough. Sorry!

File: server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java (hunk 11 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java]: I saw `"key1" ARRAY(TEXT)` in docs and supposed that on can treat it as lists are split into single items - because it's an implementation details and we could have done it in that way and I thought a small clarification is fine. Not insisting though

**Review by seut — APPROVED** [review]
👍 Thanks!

**Review by BaurzhanSakhariev — APPROVED** [review]
👍
